### ISSUE TITLE
Build fails with both SET_CUDA_HOST_COMPILER and CMAKE_CUDA_HOST_COMPILER enabled

### ISSUE THREAD

**Issue by gflegar** [ginkgo-project/ginkgo#71]
Excerpt from @venovako's e-mail:

`SET_CUDA_HOST_COMPILER=ON` and manually setting `CMAKE_CUDA_HOST_COMPILER` are mutually exclusive.
Here's what happens with the latest GitHub Ginkgo + CMake 3.11.4 if both are set:

```
Building CUDA object gpu/CMakeFiles/ginkgo_gpu.dir/matrix/coo_kernels.cu.o
cd /home/novakoni/src/GitHub/build/gpu && /state/partition1/soft/cuda-9.1/bin/nvcc -ccbin=/state/partition1/soft/gnu/gcc-6.3.0/bin/g++ -Dginkgo_gpu_EXPORTS -I/home/novakoni/src/GitHub/build/include -I/home/novakoni/src/GitHub/flx_ginkgo/include -isystem=/state/partition1/soft/cuda-9.1/include -I/home/novakoni/src/GitHub/flx_ginkgo  -Xcompiler=-mno-fma -I/home/novakoni/src/GitHub/FloatX/src --compiler-bindir /state/partition1/soft/gnu/gcc-6.3.0/bin/g++ -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_60,code=compute_60 --expt-relaxed-constexpr -O3 -DNDEBUG -Xcompiler=-fPIC   -std=c++11 -x cu -c /home/novakoni/src/GitHub/flx_ginkgo/gpu/matrix/coo_kernels.cu -o CMakeFiles/ginkgo_gpu.dir/matrix/coo_kernels.cu.o
nvcc fatal   : redefinition of argument 'compiler-bindir'
```

If you look at the compile line, it specifies `-ccbin` and again `--compiler-bindir`.
`CMAKE_CUDA_HOST_COMPILER` is manually set to `/state/partition1/soft/gnu/gcc-6.3.0/bin/g++`
The message is: if the user sets `CMAKE_CUDA_HOST_COMPILER` manually, and your system of setting the compiler also kicks in, `nvcc` dies.
It doesn't bother to check if the two definitions of the host compiler are in fact identical or not.
So, you might want to check if `CMAKE_CUDA_HOST_COMPILER` and `SET_CUDA_HOST_COMPILER` are both set and complain about it.

**Comment by gflegar** [comment]
The problem here is that [`CMAKE_CUDA_HOST_COMPILER`](https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_HOST_COMPILER.html) was introduced in CMake 3.10, and we support everything from CMake 3.8 onwards, so we cannot always rely on `CMAKE_CUDA_HOST_COMPILER` being interpreted by CMake.

My suggestion for the implementation of the `SET_CUDA_HOST_COMPILER` would be to specify the host compiler via `CMAKE_CUDA_FLAGS` for older versions of CMake (this is what we are doing now in all cases), and use `CMAKE_CUDA_HOST_COMPILER` for newer versions. This should avoid having duplicated compiler flags in newer versions of CMake.

**Comment by venovako** [comment]
One reasonable use case for someone to set `CMAKE_CUDA_HOST_COMPILER` by hand would be to explicitly use a different host compiler for `nvcc` than the one for the host-only code (e.g., because `nvcc` does not yet support the most recent compiler versions).

If the user also sets `SET_CUDA_HOST_COMPILER` to `ON` (by accident, let's say), then the compile line will contain two different host compilers - one that the user provided explicitly, and one that is chosen by `SET_CUDA_HOST_COMPILER`-related logic.

I don't know enough about CMake to tell if it is possible to create logic such as in the pseudocode below:
```
if SET_CUDA_HOST_COMPILER is ON then
    if CMAKE_CUDA_HOST_COMPILER is a non-empty string then
        warn the user about incompatible requests and die somehow
    else if CMake is recent enough then
        set CMAKE_CUDA_HOST_COMPILER to CMAKE_CXX_COMPILER
    else
        append the host-compiler-selection flag to CMAKE_CUDA_FLAGS
    end if
end if
```
That way, even with the older CMake, if a user sets `CMAKE_CUDA_HOST_COMPILER` explicitly (e.g., because he doesn't know that it is not supported), the script will warn him about (future) incompatibilities.  In such a case the user still can add the host compiler manually to `CMAKE_CUDA_FLAGS` if it is different than `CMAKE_CXX_COMPILER`.

Of course, I may be writing nonsense, but yet again, I only use CMake, don't do anything advanced with it...

**Comment by gflegar** [comment]
> I don't know enough about CMake to tell if it is possible to create logic such as in the pseudocode below:

It is possible.

The nicest thing would actually be to always just use `CMAKE_CUDA_HOST_COMPILER`, and remove `SET_CUDA_HOST_COMPILER`, no matter the version (i.e. if the CMake version is too low, we define the cache variable ourselves) - not sure if this can cause trouble later on.

**Comment by gflegar** [comment]
I'm not sure this is still relevant....
AFAIK, the CMAKE_CUDA_HOST_COMPILER flag is [available in CMake 3.8](https://github.com/Kitware/CMake/commit/489c52ce680df6439f9c1e553cd2925ca8944cb1), they just forgot to document it, so I think we removed the `SET_CUDA_HOST_COMPILER` flag in favor of the standard one.

**Comment by tcojean** [comment]
We actually still have that flag (I even updated it as part of the recent PRs). I will take care of it by simply removing our custom flag.

**Comment by gflegar** [comment]
Also, maybe mention the `CMAKE_CUDA_HOST_COMPILER_FLAG` to the README.md, and make a note that it is available on CMake 3.8.
